fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java

```diff
@@ -1112,18 +1145,29 @@ private SqlQueryData processTokenParm(QueryParameter queryParm, String tableAlia
 
                 // Include system if present.
                 if (value.getValueSystem() != null && !value.getValueSystem().isEmpty()) {
+                    Long commonTokenValueId = null;
                     if (NE.equals(operator)) {
                         whereClauseSegment.append(OR);
                     } else {
                         whereClauseSegment.append(AND);
+
+                        // use #1929 optimization if we can
+                        if (!isComposite) {
+                            commonTokenValueId = getCommonTokenValueId(value.getValueSystem(), value.getValueCode());
+                        }
                     }
+                    Integer codeSystemId = identityCache.getCodeSystemId(value.getValueSystem());
                     whereClauseSegment.append(tableAlias).append(DOT).append(CODE_SYSTEM_ID).append(operator)
-                            .append(BIND_VAR);
-
-                    codeSystemId = identityCache.getCodeSystemId(value.getValueSystem());
+                            .append(nullCheck(codeSystemId));
+
+                    if (commonTokenValueId != null) {
+                        // #1929 improves cardinality estimation
+                        // resulting in far better execution plans for many search queries. Because COMMON_TOKEN_VALUE_ID
+                        // is the primary key for the common_token_values table, we don't need the CODE_SYSTEM_ID = ? predicate.
```

Comment:
same comment as above - your comment says we don't need the CODE_SYSTEM_ID = ? predicate. Should lines 1159-1161 be moved to an else leg for the if check on line 1163?

Reply:
Same response as above

Reply:
done